CKSIncrementalStore/CKSIncrementalStoreSyncOperation.swift

```diff
@@ -639,7 +639,7 @@ class CKSIncrementalStoreSyncOperation: NSOperation {
         if self.localStoreMOC!.hasChanges
         {
             var error:NSErrorPointer = nil
-            self.localStoreMOC!.save(error)
+            self.localStoreMOC!.save(error) // TODO: saving one object at a time (or recovering and saving as much as possible) would be more robust (and catching exceptions during all the above). Otherwise a failure will result in the entire change set being lost. Also, it probably should revert/rollback on failure also.
```

Comment:
Changes coming down from iCloud aren't necessarily atomic since they could represent the combination of multiple commits, but I know what you mean.  Errors should be infrequent, but are possible.  For example, if you set up validation rules for an Integer attribute with a min and max of 0 and 1 in version 1 of your app (like to store an enum type), then later add an additional enum value and raise the max to 2.  If a user is running V1 and V2 on different devices any objects with the value 2 for the field will fail to save to V1 of the app.  Without granular saving this will halt syncing entirely, which isn't a great user experience, especially since it happens silently.  This IS going to result in exactly the inconsistency that you are trying to avoid, and a greater inconsistency than if you tried to save as much as possible.

On the subject of silent failure, it would be really useful to have a log of failures and their causes that the user can access if needed in order to troubleshoot.  This has been one of my biggest problems with using iCloud Core Data -- when it stops working you never know why.  With CloudKit there is an opportunity to provide useful information that can be communicated to technical support.